### PR TITLE
Curated Posts Test

### PR DESCRIPTION
**Pull request by tayyaba17**
Here's a (hopefully) small PR for the testing.

### CODE REVIEW

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  @Test
+  public void testFollowsOneTag() {
+    createDatastoreEntities();
+    List<String> expectedFollowedTags = new ArrayList<String>();
+    expectedFollowedTags.add("#business");
+    assertEquals(expectedFollowedTags, LoadFollowedTags.getFollowedTags("hasTags@testing.com"));
+  }
+
+  @Test
+  public void testFollowsNoTags() {
```
Comment (smarquez-jpg) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java]: Looks good, but you might want to take this out until it's filled in before you merge.

**Review by smarquez-jpg — APPROVED** [review]

**Review by smarquez-jpg — APPROVED** [review]

**Review by Andrewmob2703-zz — COMMENTED** [review]
Thanks for separating the 2. 
**_at LoginServlet 72_**
But you still want the **id next to the "UserLogin" like ("UserLogin", id) .**
That way if a user changes their nickname it's overrided in datastore and we don't store multiple nicknames for one user.
Also consider changing "UserLogin" to maybe "BlogUsers"?

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  private void createDatastoreEntities() {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Entity noTagEntity = new Entity("UserLogin");
```
Comment (martiKallas) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java]: This is lower priority, but consider creating a private helper function to add entities to Datastore. You could even make this a util class that can be used throughout testing and production code. This code is replicated in many locations. 

You could have a function like storeUserLoginEntity(id, nickname, List<tags>), and others for the other Entities.

I don't know if it is expensive to call getDatastoreService, but if it is, you can pass the DatastoreService as an argument. If the Datastore documentation doesn't say anything about the call to getDatastoreService being expensive, then it shouldn't be.

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  @Test
+  public void testFollowsNoTags() {
+    createDatastoreEntities();
+    List<String> expectedFollowedTags = new ArrayList<String>();
+    assertEquals(expectedFollowedTags, LoadFollowedTags.getFollowedTags("noTags@testing.com"));
+  }
+
+  @Test
+  public void testFollowsOneTag() {
```
Comment (martiKallas) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java:87]: I like the thorough testing, but you probably don't need this function. Does this test a different path through the code than the testFollowsMultipleTag? It's not a bad thing, but it's something you can think about in the future to save yourself a little work.

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    helper.tearDown();
+  }
+
+  private void createDatastoreEntities() {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Entity noTagEntity = new Entity("UserLogin");
+    noTagEntity.setProperty("id", "357911");
+    noTagEntity.setProperty("nickname", "noTags");
+    noTagEntity.setProperty("email", "noTags@testing.com");
```
Comment (martiKallas) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java]: As much as possible, avoid bare strings for values that you use in multiple places. For any value that you compare against in your tests, you should have a constant for it. The fields that aren't compared are fine as bare strings. Using a constant creates a clear link between the setup for the test and how it is used in the actual testing.

You should also use the BlogConstants for the properties names. If they ever change, then a person only needs to change the one file without changing your test.

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  @Test
+  public void testFollowsOneTag() {
+    createDatastoreEntities();
+    List<String> expectedFollowedTags = new ArrayList<String>();
+    expectedFollowedTags.add("#business");
+    assertEquals(expectedFollowedTags, LoadFollowedTags.getFollowedTags("hasTags@testing.com"));
+    assertEquals(1, LoadFollowedTags.getFollowedTags("hasTags@testing.com").size());
+  }
+
+    @Test
```
Comment (martiKallas) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java]: nit: indentation

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    hasMultipleTagsFollowed.add("#education");
+    hasMultipleTagsFollowed.add("#music");
+    
+    for (int i=0; i<hasMultipleTagsFollowed.size(); i++) {
+      Entity hasMultipleTagsFollowedTags = new Entity("followedTag");
+      hasMultipleTagsFollowedTags.setProperty("tag", hasMultipleTagsFollowed.get(i));
+      hasMultipleTagsFollowedTags.setProperty("email", "hasMultipleTags@testing.com");
+      datastore.put(hasMultipleTagsFollowedTags);
+    }
+
```
Comment (martiKallas) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java]: nit: don't need this blank line

**Review by martiKallas — APPROVED** [review]
I left a few comments before you submit. I like the organization and structure of this change!

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.List;
+
+
```
Comment (peizhaoou) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java]: Remove empty lines.

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+public class CuratedPostsTest {
+
```
Comment (peizhaoou) [inline portfolio/src/test/java/com/google/sps/CuratedPostsTest.java:23]: Remove extra line.

Comment (peizhaoou) [reply portfolio/src/test/java/com/google/sps/CuratedPostsTest.java:23]: Cool never mind you did that

File: portfolio/src/test/java/com/google/sps/CuratedPostsTest.java (hunk 3 above)
Comment (tayyaba17) [reply portfolio/src/test/java/com/google/sps/CuratedPostsTest.java:87]: I was going off based on the doc. I think the idea may be to catch if curated posts is working for just one tag opposed to multiple?

**Pull request merged**